### PR TITLE
Allow replacing a book's default title page

### PR DESCRIPTION
**Pull request by gtback**
Having a file named `<index>-extra-title-page.html` allows adding extra
content to a title page. This allows replacing the content entirely with
`<index>-custom-title-page.html`.

Ref: #1892

### CODE REVIEW

**Comment by gtback** [comment]
This should not create any diff in the built docs, since it's only adding new functionality. Current functionality (on the master branch of the ES reference) shouldn't be affected.

File: lib/ES/Util.pm [hunk 1]
```diff
@@ -253,21 +253,40 @@ sub build_single {
...
     $extra_title_page = $index->parent->file( $extra_title_page );
-    if ( -e $extra_title_page ) {
-        $extra_title_page = $extra_title_page->slurp( iomode => '<:encoding(UTF-8)' );
+
+    if ( -e $custom_title_page ) {
+        my $custom_contents = $custom_title_page->slurp( iomode => '<:encoding(UTF-8)' );
+        my $contents = $html_file->slurp( iomode => '<:encoding(UTF-8)' );
+        $contents =~ s|<!--START_TOC-->.*<!--END_TOC-->|$custom_contents|sm or
+            die "Couldn't set custom contents in file";
+        $html_file->spew( iomode => '>:utf8', $contents );
```
Comment (nik9000) [inline lib/ES/Util.pm:282]: It's probably worth dying if the `$extra_title_page` exists as well as the `$custom_title_page`.

Comment (gtback) [reply lib/ES/Util.pm:287]: Fixed in fe0d7f739080b267866ef3519201e6fd893daaf2

**Comment by gtback** [comment]
I added some tests and also raise an error if both the custom-title-page and extra-title-page files exist. Please take another look @olksdr and @nik9000

File: lib/ES/Util.pm [hunk 2]
```diff
@@ -253,21 +255,43 @@ sub build_single {
...
+    $custom_title_page = $index->parent->file( $custom_title_page );
+
+    (my $extra_title_page = $index_basename) =~ s/(\.x)?\.a(scii)?doc$/-extra-title-page.html/ || die;
     $extra_title_page = $index->parent->file( $extra_title_page );
-    if ( -e $extra_title_page ) {
-        $extra_title_page = $extra_title_page->slurp( iomode => '<:encoding(UTF-8)' );
+
+    if ( -e $custom_title_page ) {
+        die "Using a custom title page is incompatible with --single" if $single;
+        die "Cannot have both custom and extra title pages for for the same source file" if -e $extra_title_page;
```
Comment (olksdr) [inline lib/ES/Util.pm:281]:
```suggestion
        die "Cannot have both custom and extra title pages for the same source file" if -e $extra_title_page;
```

File: integtest/spec/single_book_spec.rb [hunk 3]
```diff
@@ -595,6 +601,67 @@ def self.setup(index_name)
...
+        repo = src.repo 'src'
+        from = repo.write 'index.adoc', INDEX_BODY
+        repo.write 'index-custom-title-page.html', '<h1>My Custom Header</h1>'
+        repo.write 'index-extra-title-page.html', '<h1>My Extra Header</h1>'
+        repo.commit 'commit outstanding'
+        dest.prepare_convert_single(from, '.').convert(expect_failure: true)
+      end
+      it 'prints an error about both files existing' do
+        expect(outputs[0]).to include(<<~LOG.strip)
+          Cannot have both custom and extra title pages for for the same source file
```
Comment (olksdr) [inline integtest/spec/single_book_spec.rb:660]:
```suggestion
          Cannot have both custom and extra title pages for the same source file
```

File: lib/ES/Util.pm (hunk 2 above)
Comment (gtback) [reply lib/ES/Util.pm:281]: 🤦 Thanks

Comment (gtback) [reply lib/ES/Util.pm]: Fixed in 3e0c59b

File: integtest/spec/single_book_spec.rb (hunk 3 above)
Comment (gtback) [reply integtest/spec/single_book_spec.rb]: Fixed in 3e0c59b

**Pull request merged**